File: TeamCode/src/main/java/org/firstinspires/ftc/common/commandbase/Command.java
package org.firstinspires.ftc.common.commandbase;

import androidx.annotation.NonNull;

public abstract class Command implements Cloneable {
	private String commandName = "";
	protected Command(String commandName) {
		this.commandName = commandName;
	}

	protected Command() { }
	/**
	 * Runs the command
	 * @return true if the command finished, false if it needs to be run again
	 */
	public abstract boolean run();
	/**
	 * Clones the command
	 * @return the cloned command
	 */
	@NonNull
	@Override
	public abstract Command clone();

	/**
	 * Gets the name of the command
	 * @return the name of the command
	 */
	public String getCommandName() {
		return commandName;
	}
}

File: TeamCode/src/main/kotlin/brickbot/quickstart/commandbase/Command.kt
package brickbot.quickstart.commandbase

abstract class Command @JvmOverloads constructor(
    protected var commandName: String = ""
) : Cloneable {
    /**
     * Runs the command
     * @return true if the command finished, false if it needs to be run again
     */
    abstract fun run(): Boolean

    /**
     * Clones the command
     * @return the cloned command
     */
    public abstract override fun clone(): Command

    /**
     * Gets the name of the command
     * @return the name of the command
     */
    fun getCommandName(): String {
        return commandName
    }
}